File: src/data/homeRows.ts
import type { Product } from "../types/product";
import { normalizeCategoryId } from "./categories";

const CHILLIES_CATEGORY_ID = "chillies-and-peppers";

/** Subtype patterns — one product per type picked first for a mixed home row. */
const CHILLI_MIX_TYPES: { id: string; re: RegExp }[] = [
  { id: "green", re: /green chilli/i },
  { id: "red", re: /red chilli/i },
  { id: "scotch", re: /scotch bonnet/i },
  { id: "habanero", re: /habanero/i },
  { id: "peri", re: /piri piri|peri-peri|peri peri/i },
  { id: "bird", re: /bird'?s eye/i },
  { id: "cayenne", re: /cayenne/i },
  { id: "thai", re: /thai chilli/i },
  { id: "jalapeno", re: /jalapeño|jalapeno/i },
  { id: "serrano", re: /serrano/i },
  { id: "dried", re: /dried.*chilli|sun-dried/i },
  { id: "sauce", re: /chilli sauce|hot pepper sauce|chilli paste|chilli powder|chilli flakes/i },
];

function isChilliProduct(p: Product): boolean {
  if (normalizeCategoryId(p.category) === CHILLIES_CATEGORY_ID) return true;
  const hay = `${p.title} ${p.description}`;
  return /chilli|chili|habanero|scotch bonnet|piri piri|peri-peri|jalapeño|jalapeno|cayenne|serrano/i.test(
    hay
  );
}

function mixChilliProducts(products: Product[], limit: number): Product[] {
  const pool = products.filter(isChilliProduct);
  const used = new Set<string>();
  const mixed: Product[] = [];

  const takeFrom = (list: Product[]) => {
    for (const p of list) {
      if (used.has(p.id)) continue;
      mixed.push(p);
      used.add(p.id);
      return true;
    }
    return false;
  };

  for (const { re } of CHILLI_MIX_TYPES) {
    if (mixed.length >= limit) break;
    const bucket = pool.filter((p) => re.test(p.title) || re.test(p.description));
    takeFrom(bucket);
  }

  let i = 0;
  while (mixed.length < limit && i < pool.length) {
    const p = pool[i++];
    if (!used.has(p.id)) {
      mixed.push(p);
      used.add(p.id);
    }
  }

  return mixed.slice(0, limit);
}

export type HomeRowConfig = {
  id: string;
  titleKey: string;
  subtitleKey: string;
  seeAllSearch?: string;
  seeAllCategoryId?: string;
  seeAllHref?: string;
  match: (product: Product) => boolean;
};

export const HOME_PRODUCT_ROWS: HomeRowConfig[] = [
  {
    id: "fresh-deals",
    titleKey: "home.rows.freshDeals",
    subtitleKey: "home.rows.freshDealsSub",
    seeAllHref: "/products?sale=1",
    match: (p) => p.originalPrice != null && p.originalPrice > p.price,
  },
  {
    id: "chillies",
    titleKey: "home.rows.chillies",
    subtitleKey: "home.rows.chilliesSub",
    seeAllCategoryId: CHILLIES_CATEGORY_ID,
    seeAllSearch: "chilli",
    match: isChilliProduct,
  },
  {
    id: "mangoes",
    titleKey: "home.rows.mangoes",
    subtitleKey: "home.rows.mangoesSub",
    seeAllSearch: "mango",
    match: (p) => /mango/i.test(p.title) || /mango/i.test(p.description),
  },
  {
    id: "chicken",
    titleKey: "home.rows.chicken",
    subtitleKey: "home.rows.chickenSub",
    seeAllSearch: "chicken",
    seeAllCategoryId: "poultry-products",
    match: (p) =>
      normalizeCategoryId(p.category) === "poultry-products" ||
      /chicken/i.test(p.title),
  },
  {
    id: "fish",
    titleKey: "home.rows.fish",
    subtitleKey: "home.rows.fishSub",
    seeAllSearch: "fish",
    seeAllCategoryId: "fish-and-aquaculture",
    match: (p) =>
      normalizeCategoryId(p.category) === "fish-and-aquaculture" ||
      /fish|tilapia|smoked|aquaculture/i.test(p.title),
  },
  {
    id: "eggs",
    titleKey: "home.rows.eggs",
    subtitleKey: "home.rows.eggsSub",
    seeAllSearch: "egg",
    seeAllCategoryId: "poultry-products",
    match: (p) => /\begg/i.test(p.title) && !/eggplant/i.test(p.title),
  },
];

function dealSavings(p: Product): number {
  if (!p.originalPrice || p.originalPrice <= p.price) return 0;
  return (p.originalPrice - p.price) / p.originalPrice;
}

export function productsForHomeRow(products: Product[], row: HomeRowConfig, limit = 12): Product[] {
  if (row.id === "chillies") {
    const mixed = mixChilliProducts(products, limit);
    if (mixed.length >= 4) return mixed;
  }

  let matched = products.filter(row.match);
  if (row.id === "fresh-deals") {
    matched = [...matched].sort((a, b) => dealSavings(b) - dealSavings(a));
  }
  if (matched.length >= 4) return matched.slice(0, limit);
  if (row.seeAllCategoryId) {
    const fromCat = products.filter(
      (p) => normalizeCategoryId(p.category) === row.seeAllCategoryId
    );
    const ids = new Set(matched.map((p) => p.id));
    for (const p of fromCat) {
      if (!ids.has(p.id)) matched.push(p);
      if (matched.length >= limit) break;
    }
  }
  return matched.slice(0, limit);
}

File: mobile/lib/data/home_rows.dart
import '../models/product.dart';
import '../utils/categories.dart';

const chilliesCategoryId = 'chillies-and-peppers';

class HomeRowConfig {
  const HomeRowConfig({
    required this.id,
    required this.match,
    this.seeAllCategoryId,
    this.seeAllSearch,
    this.seeAllHref,
    this.fallbackToCategory = true,
  });

  final String id;
  final bool Function(Product product) match;
  final String? seeAllCategoryId;
  final String? seeAllSearch;
  final String? seeAllHref;
  /// When false, never pad the row with unrelated category products.
  final bool fallbackToCategory;
}

bool Function(Product) _inCategory(String categoryId) {
  return (Product p) => normalizeCategoryId(p.category) == categoryId;
}

bool _titleMatches(Product p, String pattern) =>
    RegExp(pattern, caseSensitive: false).hasMatch(p.title);

bool _haystackMatches(Product p, String pattern) =>
    RegExp(pattern, caseSensitive: false).hasMatch('${p.title} ${p.description}');

bool _isChilliProduct(Product p) {
  if (normalizeCategoryId(p.category) == chilliesCategoryId) return true;
  return _haystackMatches(
    p,
    r'chilli|chili|habanero|scotch bonnet|piri piri|peri-peri|jalapeño|jalapeno|cayenne|serrano',
  );
}

final homeProductRows = <HomeRowConfig>[
  HomeRowConfig(
    id: 'chillies',
    seeAllCategoryId: chilliesCategoryId,
    seeAllSearch: 'chilli',
    match: _isChilliProduct,
  ),
  HomeRowConfig(
    id: 'mangoes',
    seeAllSearch: 'mango',
    match: _isMango,
  ),
  HomeRowConfig(
    id: 'bananas',
    seeAllSearch: 'banana',
    seeAllCategoryId: 'fresh-fruits',
    match: _isBanana,
  ),
  HomeRowConfig(
    id: 'fruits',
    seeAllCategoryId: 'fresh-fruits',
    seeAllSearch: 'fruit',
    match: _inCategory('fresh-fruits'),
  ),
  HomeRowConfig(
    id: 'tomatoes',
    seeAllSearch: 'tomato',
    fallbackToCategory: false,
    match: _isTomato,
  ),
  HomeRowConfig(
    id: 'vegetables',
    seeAllCategoryId: 'fresh-vegetables',
    seeAllSearch: 'vegetable',
    match: _isVegetable,
  ),
  HomeRowConfig(
    id: 'onions',
    seeAllSearch: 'onion',
    seeAllCategoryId: 'fresh-vegetables',
    match: _isOnion,
  ),
  HomeRowConfig(
    id: 'potatoes',
    seeAllCategoryId: 'roots-and-tubers',
    seeAllSearch: 'potato',
    match: _isPotato,
  ),
  HomeRowConfig(
    id: 'herbs',
    seeAllCategoryId: 'herbs-and-spices',
    seeAllSearch: 'herb',
    match: _isHerbs,
  ),
  HomeRowConfig(
    id: 'legumes',
    seeAllCategoryId: 'legumes-and-pulses',
    seeAllSearch: 'beans',
    match: _isLegumes,
  ),
  HomeRowConfig(
    id: 'nuts',
    seeAllCategoryId: 'oilseeds-and-nuts',
    seeAllSearch: 'nuts',
    match: _isNuts,
  ),
  HomeRowConfig(
    id: 'chicken',
    seeAllCategoryId: 'poultry-products',
    seeAllSearch: 'chicken',
    match: _isChicken,
  ),
  HomeRowConfig(
    id: 'beef',
    seeAllCategoryId: 'livestock-products',
    seeAllSearch: 'beef',
    match: _isBeef,
  ),
  HomeRowConfig(
    id: 'fish',
    seeAllCategoryId: 'fish-and-aquaculture',
    seeAllSearch: 'fish',
    match: _isFish,
  ),
  HomeRowConfig(
    id: 'eggs',
    seeAllCategoryId: 'poultry-products',
    seeAllSearch: 'egg',
    match: _isEgg,
  ),
  HomeRowConfig(
    id: 'dairy',
    seeAllCategoryId: 'dairy-products',
    seeAllSearch: 'milk',
    match: _inCategory('dairy-products'),
  ),
  HomeRowConfig(
    id: 'rice-grains',
    seeAllCategoryId: 'cereals-and-grains',
    seeAllSearch: 'rice',
    match: _isGrains,
  ),
  HomeRowConfig(
    id: 'coffee-tea',
    seeAllCategoryId: 'coffee-tea-cocoa',
    seeAllSearch: 'coffee',
    match: _isCoffeeTea,
  ),
];

bool _isMango(Product p) => _titleMatches(p, r'mango');

bool _isBanana(Product p) => _titleMatches(p, r'banana|plantain');

bool _isTomato(Product p) => _titleMatches(p, r'tomato');

bool _isVegetable(Product p) {
  if (_isTomato(p) || _isOnion(p)) return false;
  return normalizeCategoryId(p.category) == 'fresh-vegetables' ||
      _titleMatches(p, r'spinach|cabbage|nakati|amaranth|malakwang|greens|lettuce|cucumber');
}

bool _isOnion(Product p) => _titleMatches(p, r'onion|shallot');

bool _isPotato(Product p) =>
    normalizeCategoryId(p.category) == 'roots-and-tubers' ||
    _titleMatches(p, r'potato|sweet potato|yam|cassava|irish');

bool _isHerbs(Product p) =>
    normalizeCategoryId(p.category) == 'herbs-and-spices' ||
    _titleMatches(p, r'herb|basil|mint|coriander|parsley|rosemary|thyme');

bool _isLegumes(Product p) =>
    normalizeCategoryId(p.category) == 'legumes-and-pulses' ||
    _titleMatches(p, r'bean|pea|lentil|cowpea|groundnut');

bool _isNuts(Product p) =>
    normalizeCategoryId(p.category) == 'oilseeds-and-nuts' ||
    _titleMatches(p, r'nut|peanut|groundnut|sesame|sunflower');

bool _isChicken(Product p) =>
    !_isEgg(p) &&
    (normalizeCategoryId(p.category) == 'poultry-products' || _titleMatches(p, r'chicken'));

bool _isBeef(Product p) =>
    normalizeCategoryId(p.category) == 'livestock-products' ||
    _titleMatches(p, r'beef|lamb|goat|mutton');

bool _isFish(Product p) =>
    normalizeCategoryId(p.category) == 'fish-and-aquaculture' ||
    _titleMatches(p, r'fish|tilapia|mackerel|catfish|smoked');

bool _isEgg(Product p) =>
    _titleMatches(p, r'\begg') && !_titleMatches(p, r'eggplant');

bool _isGrains(Product p) =>
    normalizeCategoryId(p.category) == 'cereals-and-grains' ||
    _titleMatches(p, r'rice|maize|corn|wheat|flour|grain|millet');

bool _isCoffeeTea(Product p) =>
    normalizeCategoryId(p.category) == 'coffee-tea-cocoa' ||
    _titleMatches(p, r'coffee|tea|cocoa');

List<Product> productsForHomeRow(
  List<Product> products,
  HomeRowConfig row, {
  int limit = 12,
}) {
  var matched = products.where((p) => p.inStock && row.match(p)).toList();

  if (matched.length >= 4) return matched.take(limit).toList();

  if (row.fallbackToCategory && row.seeAllCategoryId != null) {
    final fromCat = products
        .where((p) => p.inStock && normalizeCategoryId(p.category) == row.seeAllCategoryId)
        .toList();
    final ids = matched.map((p) => p.id).toSet();
    for (final p in fromCat) {
      if (row.match(p) && ids.add(p.id)) matched.add(p);
      if (matched.length >= limit) break;
    }
  }

  return matched.take(limit).toList();
}

String? homeRowSeeAllPath(HomeRowConfig row) {
  if (row.seeAllHref != null) return row.seeAllHref;
  if (row.seeAllCategoryId != null) return '/category/${row.seeAllCategoryId}';
  if (row.seeAllSearch != null) return '/search?q=${Uri.encodeComponent(row.seeAllSearch!)}';
  return null;
}
